hibernate-core/src/main/java/org/hibernate/id/enhanced/HiLoOptimizer.java

```diff
@@ -53,12 +53,78 @@
 public class HiLoOptimizer extends AbstractOptimizer {
 	private static final Logger log = Logger.getLogger( HiLoOptimizer.class );
 
-	private static class GenerationState {
+	protected static class GenerationState {
+		private final int incrementSize;
 		private IntegralDataTypeHolder lastSourceValue;
 		private IntegralDataTypeHolder upperLimit;
 		private IntegralDataTypeHolder value;
-	}
 
+		/**
+		 * @param incrementSize - the increment size.
+		 * @throws HibernateException if {@code incrementSize < 1}.
+		 */
+		private GenerationState(int incrementSize) {
+			if ( incrementSize < 1 ) {
+				throw new HibernateException( "increment size cannot be less than 1" );
+			}
+			this.incrementSize = incrementSize;
+		}
+
+		/**
+		 * Has this {@link GenerationState} been initialized with a
+		 * "source" value (e.g., from the database).
+		 *
+		 * @return true, if it is initialized; false, otherwise.
+		 */
+		public final boolean isInitialized() {
+			return lastSourceValue != null;
+		}
+
+		/**
+		 * Is a new "source" value (e.g., from the database) required
+		 * in order to obtain the next value? A new source value is
+		 * required when {@link #isInitialized()} is {@code false},
+		 * or the current value has already reached the upper limit.
+		 * When {@code true} is returned, {@link #updateFromNewSourceValue(IntegralDataTypeHolder)}
+		 * must be called in order to obtain the next value.
+		 *
+		 * @return true, if a new source value is required in order
+		 *               to obtain the next value; false, otherwise.
+		 */
+		public final boolean requiresNewSourceValue() {
+			return ! isInitialized() || ! upperLimit.gt( value );
+		}
+
+		/**
+		 * Returns the next value, and increments the internal value
+		 * in preparation for the following value to be requested.
+		 *
+		 * @return the next value.
+		 * @throws HibernateException if {@link #requiresNewSourceValue()} is {@code true}.
+		 */
+		public final Number makeValueThenIncrement() {
+			if ( requiresNewSourceValue() ) {
+				throw new HibernateException( "Value is either uninitialized or has already reached the upper limit. GenerationState#updateFromNewSourceValue must be called before obtaining the next value." );
+			}
+			return value.makeValueThenIncrement();
+		}
+
+		/**
+		 * Update this GenerationState with a new source value (e.g., from the database).
+		 *
+		 * @param sourceValue - the value from the "source" of values (e.g., from the database);
+		 * @throws HibernateException if sourceValue < 1.
+		 */
+		public final void updateFromNewSourceValue(IntegralDataTypeHolder sourceValue) {
+			// Copy sourceValue, so caller cannot modify the internal state.
+			lastSourceValue = sourceValue.copy();
+
+			// upperLimit defines the upper end of the bucket values
+			upperLimit = lastSourceValue.copy().multiplyBy( incrementSize ).increment();
```

Comment:
may be I'm missing something but cannot just be `upperLimit = sourceValue.copy().multiplyBy( incrementSize ).increment();` removing the ` lastSourceValue = sourceValue.copy();` ?